Drastically simplify distribution building for 2.11.

adriaanm opened the pull request:
See README and individual commits for info.

The main idea is that these scripts should do nothing more than repackaging org.scala-lang % scala-dist and its transitive dependencies. No duplication allowed. The process is reduced to pushing a tag to scala/scala-dist by the same name as the one from scala/scala (whose artifacts have been published to maven), and running `sbt s3-upload` on windows (for the msi) and unix (everything else).

We no longer package source code ourselves. GitHub provides source archives for every tag.

Review by @jsuereth, /cc @retronym, @gkossakowski.

This has all been tested locally. I'll do a dry run on jenkins tomorrow and link to the results.

adriaanm committed:
Delete cruft, simplify directory structure.

adriaanm committed:
Wix.settings: mappings for creating the MSI.

Takes the result of staging universal and universal-docs,
and creates a WIX xml description for `packageBin in Windows`
to be able to create an MSI.

We no longer distributing examples, and link to GitHub for sources.

As .bat files are published to maven using the windows line endings,
no need to convert.

PS: Would've called in Windows.settings, but that clashes with the
packager's Windows config object.

adriaanm committed:
Generic.settings: mappings for bin/, lib/, doc/ and man/

A Scala distribution is fully determined by the contents and dependencies of
the maven artifact "org.scala-lang" % "scala-dist" % version.value.

The contents of scala-dist itself is extracted to populate bin/, doc/ and man/.
The (transitive) dependencies populate the lib/ folder.

This configures sbt native packager's Universal config.

The next commit will take the jars under the javadoc classifier and extract/move
them to api/ to provide settings for the packager's UniversalDocs config.

adriaanm committed:
Docs.settings: mappings for the api/ directory.

All dependencies of scala-dist under the javadoc classifier are included as
"api/jars/$artifactId-$version-javadoc.jar".

The core jars are also expanded under api/scala-{library|reflect|compiler}.

Also make constructing mappings reusable.

Needed to use updateClassifiers to get -javadoc artifacts,
don't seem to need to include the javadoc classifier in the scala-dist config.

adriaanm committed:
Commented stuff carried over from old version.

adriaanm committed:
Remove duplication in license/copyright.

adriaanm committed:
Derive version from git tag.

The version of this build determines the Scala version to package.
We look at the closest git tag that matches v[0-9].* to derive it.
For testing, the version may be overridden with -Dproject.version=...

adriaanm committed:
Create debian & rpm packages.

`debian:packageBin` and `rpm:packageBin` create `scala-$version.[deb|rpm]` in `target/`

- subdirs not requiring special treatment are rooted under `installTargetUnix.value`
- docs go to `installTargetUnixDocs.value`
- create symlinks under /usr/bin/ for all scripts in `installTargetUnix.value / "bin"`
- make scripts in `installTargetUnix.value / "bin"` executable, drop ".bat" files
- map man to /usr/share/man/
- map doc to `installTargetUnixDocs.value`
- map api to `installTargetUnixDocs.value / "api"`

adriaanm committed:
Standardize msi's name.

adriaanm committed:
Discontinue packaging of tool-support.

Hopefully someone will take over maintenance, or we will when demand arises.

adriaanm committed:
s3-upload support, and cleanup

Only load Wix settings when on Windows,
do the rest of the work when not running on windows
(packageBin in universal, universal-docs, debian, rpm)

adriaanm committed:
Update README.

adriaanm commented:
This depends on scala/scala#3277 -- UPDATE: this has been merged, and jobs are successful, so full steam ahead!

adriaanm committed:
Split out versioning.

The windows version is used for the unix versions,
so can't have its computation only run when running on windows.

adriaanm commented:
Works on unix: https://scala-webapps.epfl.ch/jenkins/view/scala-release-2.11.x/job/scala-release-2.11.x-unix/10/console
Works on windows: https://scala-webapps.epfl.ch/jenkins/view/scala-release-2.11.x/job/scala-release-2.11.x-windows/6/console (the failure is an S3 failure also sometimes seen in the old jobs because of VirtualBox issues -- retrying as job 7)

adriaanm commented:
The last thing between us and one-click releases for both maven and native packages is integrating sonatype staging repo closing (https://github.com/adriaanm/binfu/blob/master/sonafu.sh) into https://scala-webapps.epfl.ch/jenkins/view/scala-release-2.11.x/job/scala-release-2.11.x/.

For tying the binary compatibility knot (https://github.com/scala/scala-modules-build/blob/master/build.sh) I'll need to implement something very similar, so I'm hopeful we'll actually get there!

Of course, there's always more to be done. Then, we'll need automation for generating scala/scala-lang PRs (release notes, download pages,...) and a big red RELEASE button (which would release the closed sonatype repo).

jsuereth commented on README.md:
```diff
@@ -1,44 +1,36 @@
-  * Start [sbt](https://github.com/harrah/xsbt) in the `<scala/scala-dist>` directory.
-  * Run one of the build commands.
+Due to limited resources, the native packages are quite rudimental,
+and tool-support isn't packaged at all.
```
That's probably ok.  maybe we can get some community contributions after this whole process is dramatically simplified.

jsuereth commented on build.sbt:
```diff
@@ -0,0 +1,28 @@
+// The version of this build determines the Scala version to package.
+// We look at the closest git tag that matches v[0-9].* to derive it.
+// For testing, the version may be overridden with -Dproject.version=...
+versionWithGit
```
I always smile when I see people using this.

adriaanm replied:
:-) git ftw! I think we'll see tag-driven scala releases for 2.11.x!

jsuereth commented on project/Docs.scala:
```diff
@@ -0,0 +1,44 @@
+    name in UniversalDocs := s"scala-docs-${version.value}",
+    // libraryDependencies += scalaDistDep(version.value, "javadoc"), // seems not to be necessary
+    // need updateClassifiers to get javadoc jars
+    mappings in UniversalDocs ++= createMappingsWith(updateClassifiers.value.toSeq, universalDocsMappings)
```
nice...

jsuereth commented on project/ScalaDist.scala:
```diff
@@ -0,0 +1,94 @@
+  // used to make s3-upload upload the file produced by fileTask to the path scala/$version/${file.name}
+  private def uploadMapping(fileTask: TaskKey[File]) = Def.task {
+    val file = fileTask.value
+    file -> s"scala/${version.value}/${file.getName}"
```
nice simplification!

jsuereth commented on project/Unix.scala:
```diff
@@ -0,0 +1,107 @@
+    installTargetUnix     := file("/usr/share/scala"),
+    installTargetUnixDocs := file("/usr/share/doc/scala"),
+
+    (packageBin in Rpm) <<= ((packageBin in Rpm)
```
Do you need it it to be staged?  I thought if you just et up the mappings it should work.

adriaanm replied:
good point -- remnant from when we were still patching things; trying without and seems to going through
I'll add a commit

jsuereth commented on project/Unix.scala:
```diff
@@ -0,0 +1,107 @@
+    linuxPackageSymlinks ++= (
+      (mappings in Universal).value collect {
+        case (file, name) if (name startsWith "bin/") && !(name endsWith ".bat") =>
+          LinuxSymlink("/usr/" + name, (installTargetUnix.value / name).getAbsolutePath)
```
nice.  Is this borrowed from native packager?  Did the "GenericMappingsToLinux" settings not work?

adriaanm replied:
It is borrowed -- sorry, hacked the last bits in anger, so didn't reuse all the bits I could've reused. Let's gradually refine from here.

jsuereth commented on project/Unix.scala:
```diff
@@ -0,0 +1,107 @@
+
+      val rpm = (packageBin in Rpm).value match {
+        case reported if reported.exists => reported
+        case _ => // hack on top of hack because RpmHelper.buildRpm is broken -- `spec.meta.arch` doesn't necessarily match the arch `rpmbuild` decided on
```
This seems to be a MacOSX issue.  IF you think you know the issue, let me know. I'd love to fix it, but wasn't able to isolate the problem.  It isn't borked on RedHat/Ubuntu.

adriaanm replied:
clarified mac issue in comment

jsuereth commented on project/Unix.scala:
```diff
@@ -0,0 +1,107 @@
+    // Debian Specific
+    name in Debian    := "scala",
+    debianPackageDependencies += "openjdk-6-jre | java6-runtime",
+    debianPackageDependencies += "libjansi-java",
```
not really needed anymore.

jsuereth commented on project/Versioning.scala:
```diff
@@ -0,0 +1,50 @@
+
+    version match {
+      case Majors(major, minor, bugfix, rest) => Option(rest) getOrElse "" match {
+        case Milestones(null, num)            => major + "." + minor + "." + bugfix + "." + calculateNumberFour(0,0,num.toInt)
```
Yay for ensuring windows upgrades work correctly.

jsuereth commented on project/Wix.scala:
```diff
@@ -0,0 +1,108 @@
+    // distributionFiles in Windows += (packageMsi in Windows).value,
+
+    wixProductId        := "7606e6da-e168-42b5-8345-b08bf774cb30",
+    wixProductUpgradeId := "6061c134-67c7-4fb2-aff5-32b01a186968",
```
These are the same as before, right?

adriaanm replied:
yep

jsuereth commented:
This looks very very good man!   I have one comment, one question, and otherwise looks great from me.

adriaanm commented:
Thanks!

adriaanm committed:
Review feedback. Unix mappings don't depend on stage.

Because we no longer need to patch the scripts (as we gave Scala proper home,
rather than plopping all the jars in /usr/share/java).

adriaanm commented:
@jsuereth, I believe I've addressed your feedback!

adriaanm committed:
Clean up windows version computation.
